Build error: cannot call member function ‘void taco::Iterators::createAccessIterators(taco::Access, taco::Format, taco::ir::Expr)’ without object

I see the following error when building on Ubuntu 16.04

```
[  3%] Building CXX object src/CMakeFiles/taco.dir/lower/iterator.cpp.o
/home/riley/compilers/proj/rjw2624_hh26833_compilers_final_proj/taco/src/lower/iterator.cpp: In instantiation of ‘taco::Iterators::Iterators(taco::IndexStmt, const std::map<taco::TensorVar, taco::ir::Expr>&)::<lambda(auto:3)> [with auto:3 = const taco::AccessNode*]’:
/usr/include/c++/5/type_traits:2346:26:   required by substitution of ‘template<class _Fn, class ... _Args> static std::__success_type<decltype (declval<_Fn>()((declval<_Args>)()...))> std::__result_of_other_impl::_S_test(int) [with _Fn = taco::Iterators::Iterators(taco::IndexStmt, const std::map<taco::TensorVar, taco::ir::Expr>&)::<lambda(auto:3)>&; _Args = {const taco::AccessNode*}]’
/usr/include/c++/5/type_traits:2357:55:   required from ‘struct std::__result_of_impl<false, false, taco::Iterators::Iterators(taco::IndexStmt, const std::map<taco::TensorVar, taco::ir::Expr>&)::<lambda(auto:3)>&, const taco::AccessNode*>’
/usr/include/c++/5/type_traits:2361:12:   required from ‘class std::result_of<taco::Iterators::Iterators(taco::IndexStmt, const std::map<taco::TensorVar, taco::ir::Expr>&)::<lambda(auto:3)>&(const taco::AccessNode*)>’
/usr/include/c++/5/functional:2053:9:   required by substitution of ‘template<class _Functor, class, class> std::function<_Res(_ArgTypes ...)>::function(_Functor) [with _Functor = taco::Iterators::Iterators(taco::IndexStmt, const std::map<taco::TensorVar, taco::ir::Expr>&)::<lambda(auto:3)>; <template-parameter-1-2> = void; <template-parameter-1-3> = <missing>]’
/home/riley/compilers/proj/rjw2624_hh26833_compilers_final_proj/taco/src/lower/iterator.cpp:393:6:   required from here
/home/riley/compilers/proj/rjw2624_hh26833_compilers_final_proj/taco/src/lower/iterator.cpp:392:28: error: cannot call member function ‘void taco::Iterators::createAccessIterators(taco::Access, taco::Format, taco::ir::Expr)’ without object
       createAccessIterators(Access(n), format, tensorIR);
                            ^
src/CMakeFiles/taco.dir/build.make:638: recipe for target 'src/CMakeFiles/taco.dir/lower/iterator.cpp.o' failed
make[2]: *** [src/CMakeFiles/taco.dir/lower/iterator.cpp.o] Error 1
CMakeFiles/Makefile2:117: recipe for target 'src/CMakeFiles/taco.dir/all' failed
make[1]: *** [src/CMakeFiles/taco.dir/all] Error 2
Makefile:127: recipe for target 'all' failed
make: *** [all] Error 2
```

I've got g++ 5.5.0 installed:
g++ (Ubuntu 5.5.0-12ubuntu1~16.04) 5.5.0 20171010

I was able to successfully build on a different machine running Ubuntu 18.04. Does anyone know the cause of this error?

@stephenchouca Thanks for your reply. Add `this->` resolves the error but new errors occur.

```
/home/dongxiao/working/taco-new-scheling-lang/test/tests-scheduling.cpp: In member function ‘virtual void scheduling_lowerSparseCopy_Test::TestBody()’:
/home/dongxiao/working/taco-new-scheling-lang/test/tests-scheduling.cpp:116:38: error: call of overloaded ‘Tensor(const char [2], <brace-enclosed initializer list>, <brace-enclosed initializer list>)’ is ambiguous
   Tensor<double> A("A", {8}, {Sparse});
                                      ^
In file included from /home/dongxiao/working/taco-new-scheling-lang/test/test.h:11:0,
                 from /home/dongxiao/working/taco-new-scheling-lang/test/tests-scheduling.cpp:3:
/home/dongxiao/working/taco-new-scheling-lang/include/taco/tensor.h:513:3: note: candidate: taco::Tensor<CType>::Tensor(std::__cxx11::string, std::vector<int>, taco::Format) [with CType = double; std::__cxx11::string = std::__cxx11::basic_string<char>]
   Tensor(std::string name, std::vector<int> dimensions, Format format)
   ^
/home/dongxiao/working/taco-new-scheling-lang/include/taco/tensor.h:508:3: note: candidate: taco::Tensor<CType>::Tensor(std::__cxx11::string, std::vector<int>, taco::ModeFormat) [with CType = double; std::__cxx11::string = std::__cxx11::basic_string<char>]
   Tensor(std::string name, std::vector<int> dimensions, 

```

--- a/src/lower/iterator.cpp
+++ b/src/lower/iterator.cpp
@@ -417,7 +417,7 @@ Iterators::Iterators(IndexStmt stmt, const map<TensorVar, Expr>& tensorVars)
       taco_iassert(util::contains(tensorVars, n->tensorVar));
       Expr tensorIR = tensorVars.at(n->tensorVar);
       Format format = n->tensorVar.getFormat();
-      createAccessIterators(Access(n), format, tensorIR, provGraph);
+      this->createAccessIterators(Access(n), format, tensorIR, provGraph);
     }),
     function<void(const AssignmentNode*, Matcher*)>([&](auto n, auto m) {
       m->match(n->rhs);

--- a/test/tests-scheduling-eval.cpp
+++ b/test/tests-scheduling-eval.cpp
@@ -323,8 +323,8 @@ TEST(scheduling_eval, test_spmvCPU_temp) {
   int NUM_J = 1039/10;
   float SPARSITY = .3;
   Tensor<double> A("A", {NUM_I, NUM_J}, CSR);
-  Tensor<double> x("x", {NUM_J}, {Dense});
-  Tensor<double> y("y", {NUM_I}, {Dense});
+  Tensor<double> x("x", {NUM_J}, Format({Dense}));
+  Tensor<double> y("y", {NUM_I}, Format({Dense}));
 
   srand(4353);
   for (int i = 0; i < NUM_I; i++) {
@@ -362,7 +362,7 @@ TEST(scheduling_eval, test_spmvCPU_temp) {
   y.assemble();
   y.compute();
 
-  Tensor<double> expected("expected", {NUM_I}, {Dense});
+  Tensor<double> expected("expected", {NUM_I}, Format({Dense}));
   expected(i) = A(i, j) * x(j);
   expected.compile();
   expected.assemble();
@@ -379,8 +379,8 @@ TEST(scheduling_eval, example_spmvCPU_splitpos) {
   float SPARSITY = .3;
   int CHUNK_SIZE = 16;
   Tensor<double> A("A", {NUM_I, NUM_J}, CSR);
-  Tensor<double> x("x", {NUM_J}, {Dense});
-  Tensor<double> y("y", {NUM_I}, {Dense});
+  Tensor<double> x("x", {NUM_J}, Format({Dense}));
+  Tensor<double> y("y", {NUM_I}, Format({Dense}));
 
   srand(53535);
   for (int i = 0; i < NUM_I; i++) {
@@ -415,7 +415,7 @@ TEST(scheduling_eval, example_spmvCPU_splitpos) {
   y.assemble();
   y.compute();
 
-  Tensor<double> expected("expected", {NUM_I}, {Dense});
+  Tensor<double> expected("expected", {NUM_I}, Format({Dense}));
   expected(i) = A(i, j) * x(j);
   expected.compile();
   expected.assemble();
@@ -542,8 +542,8 @@ TEST(scheduling_eval, spmvCPU) {
   int NUM_J = 1039/10;
   float SPARSITY = .3;
   Tensor<double> A("A", {NUM_I, NUM_J}, CSR);
-  Tensor<double> x("x", {NUM_J}, {Dense});
-  Tensor<double> y("y", {NUM_I}, {Dense});
+  Tensor<double> x("x", {NUM_J}, Format({Dense}));
+  Tensor<double> y("y", {NUM_I}, Format({Dense}));
 
   srand(120);
   for (int i = 0; i < NUM_I; i++) {
@@ -574,7 +574,7 @@ TEST(scheduling_eval, spmvCPU) {
   y.assemble();
   y.compute();
 
-  Tensor<double> expected("expected", {NUM_I}, {Dense});
+  Tensor<double> expected("expected", {NUM_I}, Format({Dense}));
   expected(i) = A(i, j) * x(j);
   expected.compile();
   expected.assemble();
@@ -592,7 +592,7 @@ TEST(scheduling_eval, ttvCPU) {
   float SPARSITY = .3;
   Tensor<double> A("A", {NUM_I, NUM_J}, {Dense, Dense}); // TODO: change to sparse outputs
   Tensor<double> B("B", {NUM_I, NUM_J, NUM_K}, {Sparse, Sparse, Sparse});
-  Tensor<double> c("c", {NUM_K}, {Dense});
+  Tensor<double> c("c", {NUM_K}, Format({Dense}));
 
   srand(9536);
   for (int i = 0; i < NUM_I; i++) {
@@ -757,8 +757,8 @@ TEST(scheduling_eval, spmvGPU) {
   int NUM_J = 1039/10;
   float SPARSITY = .01;
   Tensor<double> A("A", {NUM_I, NUM_J}, CSR);
-  Tensor<double> x("x", {NUM_J}, {Dense});
-  Tensor<double> y("y", {NUM_I}, {Dense});
+  Tensor<double> x("x", {NUM_J}, Format({Dense}));
+  Tensor<double> y("y", {NUM_I}, Format({Dense}));
 
   srand(94353);
   for (int i = 0; i < NUM_I; i++) {
@@ -789,7 +789,7 @@ TEST(scheduling_eval, spmvGPU) {
   y.assemble();
   y.compute();
 
-  Tensor<double> expected("expected", {NUM_I}, {Dense});
+  Tensor<double> expected("expected", {NUM_I}, Format({Dense}));
   expected(i) = A(i, j) * x(j);
   expected.compile();
   expected.assemble();
@@ -1023,7 +1023,7 @@ TEST(scheduling_eval, ttvGPU) {
   float SPARSITY = .3;
   Tensor<double> A("A", {NUM_I, NUM_J}, {Dense, Dense}); // TODO: change to sparse outputs
   Tensor<double> B("B", {NUM_I, NUM_J, NUM_K}, {Sparse, Sparse, Sparse});
-  Tensor<double> c("c", {NUM_K}, {Dense});
+  Tensor<double> c("c", {NUM_K}, Format({Dense}));
 
   srand(35325);
   for (int i = 0; i < NUM_I; i++) {
@@ -1169,8 +1169,8 @@ TEST(generate_evaluation_files, DISABLED_cpu) {
     stringstream source;
     std::shared_ptr<ir::CodeGen> codegen = ir::CodeGen::init_default(source, ir::CodeGen::ImplementationGen);
     Tensor<double> A("A", {NUM_I, NUM_J}, CSR);
-    Tensor<double> x("x", {NUM_J}, {Dense});
-    Tensor<double> y("y", {NUM_I}, {Dense});
+    Tensor<double> x("x", {NUM_J}, Format({Dense}));
+    Tensor<double> y("y", {NUM_I}, Format({Dense}));
     y(i) = A(i, j) * x(j);
     IndexStmt stmt = y.getAssignment().concretize();
     bool isFirst = true;
@@ -1237,7 +1237,7 @@ TEST(generate_evaluation_files, DISABLED_cpu) {
     std::shared_ptr<ir::CodeGen> codegen = ir::CodeGen::init_default(source, ir::CodeGen::ImplementationGen);
     Tensor<double> A("A", {NUM_I, NUM_J}, {Dense, Dense}); // TODO: change to sparse outputs
     Tensor<double> B("B", {NUM_I, NUM_J, NUM_K}, {Sparse, Sparse, Sparse});
-    Tensor<double> c("c", {NUM_K}, {Dense});
+    Tensor<double> c("c", {NUM_K}, Format({Dense}));
     A(i,j) = B(i,j,k) * c(k);
     IndexStmt stmt = A.getAssignment().concretize();
     bool isFirst = true;
@@ -1481,8 +1481,8 @@ TEST(generate_evaluation_files, DISABLED_gpu) {
     stringstream source;
     std::shared_ptr<ir::CodeGen> codegen = ir::CodeGen::init_default(source, ir::CodeGen::ImplementationGen);
     Tensor<double> A("A", {NUM_I, NUM_J}, CSR);
-    Tensor<double> x("x", {NUM_J}, {Dense});
-    Tensor<double> y("y", {NUM_I}, {Dense});
+    Tensor<double> x("x", {NUM_J}, Format({Dense}));
+    Tensor<double> y("y", {NUM_I}, Format({Dense}));
     IndexExpr precomputed = A(i, j) * x(j);
     y(i) = precomputed;
     IndexStmt stmt = y.getAssignment().concretize();
@@ -1509,8 +1509,8 @@ TEST(generate_evaluation_files, DISABLED_gpu) {
     stringstream source;
     std::shared_ptr<ir::CodeGen> codegen = ir::CodeGen::init_default(source, ir::CodeGen::ImplementationGen);
     Tensor<double> A("A", {NUM_I, NUM_J}, CSR);
-    Tensor<double> x("x", {NUM_J}, {Dense});
-    Tensor<double> y("y", {NUM_I}, {Dense});
+    Tensor<double> x("x", {NUM_J}, Format({Dense}));
+    Tensor<double> y("y", {NUM_I}, Format({Dense}));
     IndexExpr precomputed = A(i, j) * x(j);
     y(i) = precomputed;
     IndexStmt stmt = y.getAssignment().concretize();
@@ -1582,7 +1582,7 @@ TEST(generate_evaluation_files, DISABLED_gpu) {
     std::shared_ptr<ir::CodeGen> codegen = ir::CodeGen::init_default(source, ir::CodeGen::ImplementationGen);
     Tensor<double> A("A", {NUM_I, NUM_J}, {Dense, Dense}); // TODO: change to sparse outputs
     Tensor<double> B("B", {NUM_I, NUM_J, NUM_K}, {Sparse, Sparse, Sparse});
-    Tensor<double> c("c", {NUM_K}, {Dense});
+    Tensor<double> c("c", {NUM_K}, Format({Dense}));
     IndexExpr precomputedExpr = B(i,j,k) * c(k);
     A(i,j) = precomputedExpr;
     IndexStmt stmt = A.getAssignment().concretize();
@@ -1665,8 +1665,8 @@ TEST(generate_figures, DISABLED_cpu) {
     stringstream source;
     std::shared_ptr<ir::CodeGen> codegen = ir::CodeGen::init_default(source, ir::CodeGen::ImplementationGen);
     Tensor<double> A("A", {NUM_I, NUM_J}, CSR);
-    Tensor<double> x("x", {NUM_J}, {Dense});
-    Tensor<double> y("y", {NUM_I}, {Dense});
+    Tensor<double> x("x", {NUM_J}, Format({Dense}));
+    Tensor<double> y("y", {NUM_I}, Format({Dense}));
     y(i) = A(i, j) * x(j);
     IndexStmt stmt = y.getAssignment().concretize();
     bool isFirst = true;

--- a/test/tests-scheduling.cpp
+++ b/test/tests-scheduling.cpp
@@ -114,8 +114,8 @@ TEST(scheduling, lowerDenseMatrixMul) {
 }
 
 TEST(scheduling, lowerSparseCopy) {
-  Tensor<double> A("A", {8}, {Sparse});
-  Tensor<double> C("C", {8}, {Dense});
+  Tensor<double> A("A", {8}, Format({Sparse}));
+  Tensor<double> C("C", {8}, Format({Dense}));
 
   for (int i = 0; i < 8; i++) {
     if (i % 2 == 0) {
@@ -136,7 +136,7 @@ TEST(scheduling, lowerSparseCopy) {
   C.assemble();
   C.compute();
 
-  Tensor<double> expected("expected", {8}, {Dense});
+  Tensor<double> expected("expected", {8}, Format({Dense}));
   expected(i) = A(i);
   expected.compile();
   expected.assemble();
@@ -149,9 +149,9 @@ TEST(scheduling, lowerSparseCopy) {
 }
 
 TEST(scheduling, lowerSparseMulDense) {
-  Tensor<double> A("A", {8}, {Sparse});
-  Tensor<double> B("B", {8}, {Dense});
-  Tensor<double> C("C", {8}, {Dense});
+  Tensor<double> A("A", {8}, Format({Sparse}));
+  Tensor<double> B("B", {8}, Format({Dense}));
+  Tensor<double> C("C", {8}, Format({Dense}));
 
   for (int i = 0; i < 8; i++) {
     if (i % 2 == 0) {
@@ -174,7 +174,7 @@ TEST(scheduling, lowerSparseMulDense) {
   C.assemble();
   C.compute();
 
-  Tensor<double> expected("expected", {8}, {Dense});
+  Tensor<double> expected("expected", {8}, Format({Dense}));
   expected(i) = A(i) * B(i);
   expected.compile();
   expected.assemble();
@@ -187,9 +187,9 @@ TEST(scheduling, lowerSparseMulDense) {
 }
 
 TEST(scheduling, lowerSparseMulSparse) {
-  Tensor<double> A("A", {8}, {Sparse});
-  Tensor<double> B("B", {8}, {Sparse});
-  Tensor<double> C("C", {8}, {Dense});
+  Tensor<double> A("A", {8}, Format({Sparse}));
+  Tensor<double> B("B", {8}, Format({Sparse}));
+  Tensor<double> C("C", {8}, Format({Dense}));
 
   for (int i = 0; i < 8; i++) {
     if (i % 2 == 0) {
@@ -214,7 +214,7 @@ TEST(scheduling, lowerSparseMulSparse) {
   C.assemble();
   C.compute();
 
-  Tensor<double> expected("expected", {8}, {Dense});
+  Tensor<double> expected("expected", {8}, Format({Dense}));
   expected(i) = A(i) * B(i);
   expected.compile();
   expected.assemble();
@@ -227,9 +227,9 @@ TEST(scheduling, lowerSparseMulSparse) {
 }
 
 TEST(scheduling, lowerSparseAddSparse) {
-  Tensor<double> A("A", {8}, {Sparse});
-  Tensor<double> B("B", {8}, {Sparse});
-  Tensor<double> C("C", {8}, {Dense});
+  Tensor<double> A("A", {8}, Format({Sparse}));
+  Tensor<double> B("B", {8}, Format({Sparse}));
+  Tensor<double> C("C", {8}, Format({Dense}));
 
   for (int i = 0; i < 8; i++) {
     if (i % 2 == 0) {
@@ -254,7 +254,7 @@ TEST(scheduling, lowerSparseAddSparse) {
   C.assemble();
   C.compute();
 
-  Tensor<double> expected("expected", {8}, {Dense});
+  Tensor<double> expected("expected", {8}, Format({Dense}));
   expected(i) = A(i) + B(i);
   expected.compile();
   expected.assemble();
@@ -270,7 +270,7 @@ TEST(scheduling, lowerSparseAddSparse) {
 TEST(scheduling, lowerSparseMatrixMul) {
   Tensor<double> A("A", {8, 8}, CSR);
   Tensor<double> B("B", {8, 8}, CSC);
-  Tensor<double> C("C", {8, 8}, {Dense, Dense});
+  Tensor<double> C("C", {8, 8}, Format({Dense, Dense}));
 
   for (int i = 0; i < 8; i++) {
     for (int j = 0; j < 8; j++) {
@@ -318,8 +318,8 @@ TEST(scheduling, lowerSparseMatrixMul) {
 }
 
 TEST(scheduling, parallelizeAtomicReduction) {
-  Tensor<double> A("A", {8}, {Sparse});
-  Tensor<double> B("B", {8}, {Dense});
+  Tensor<double> A("A", {8}, Format({Sparse}));
+  Tensor<double> B("B", {8}, Format({Dense}));
   Tensor<double> C("C");
 
   for (int i = 0; i < 8; i++) {
@@ -365,8 +365,8 @@ TEST(scheduling, parallelizeAtomicReduction) {
 }
 
 TEST(scheduling, parallelizeTemporaryReduction) {
-  Tensor<double> A("A", {8}, {Sparse});
-  Tensor<double> B("B", {8}, {Dense});
+  Tensor<double> A("A", {8}, Format({Sparse}));
+  Tensor<double> B("B", {8}, Format({Dense}));
   Tensor<double> C("C");
 
   for (int i = 0; i < 8; i++) {
@@ -412,9 +412,9 @@ TEST(scheduling, parallelizeTemporaryReduction) {
 }
 
 TEST(scheduling, multilevel_tiling) {
-  Tensor<double> A("A", {8}, {Sparse});
-  Tensor<double> B("B", {8}, {Sparse});
-  Tensor<double> C("C", {8}, {Dense});
+  Tensor<double> A("A", {8}, Format({Sparse}));
+  Tensor<double> B("B", {8}, Format({Sparse}));
+  Tensor<double> C("C", {8}, Format({Dense}));
 
   for (int i = 0; i < 8; i++) {
     A.insert({i}, (double) i);
@@ -426,7 +426,7 @@ TEST(scheduling, multilevel_tiling) {
   A.pack();
   B.pack();
 
-  Tensor<double> expected("expected", {8}, {Dense});
+  Tensor<double> expected("expected", {8}, Format({Dense}));
   expected(i) = A(i) * B(i);
   expected.compile();
   expected.assemble();
@@ -474,7 +474,7 @@ TEST(scheduling, multilevel_tiling) {
 }
 
 TEST(scheduling, pos_noop) {
-  Tensor<double> A("A", {8}, {Sparse});
+  Tensor<double> A("A", {8}, Format({Sparse}));
   Tensor<double> C("C");
 
   for (int i = 0; i < 8; i++) {
@@ -508,9 +508,9 @@ TEST(scheduling, pos_noop) {
 }
 
 TEST(scheduling, pos_mul_dense) {
-  Tensor<double> A("A", {8}, {Sparse});
-  Tensor<double> B("B", {8}, {Dense});
-  Tensor<double> C("C", {8}, {Dense});
+  Tensor<double> A("A", {8}, Format({Sparse}));
+  Tensor<double> B("B", {8}, Format({Dense}));
+  Tensor<double> C("C", {8}, Format({Dense}));
 
   for (int i = 0; i < 8; i++) {
     if (i % 2 == 0) {
@@ -536,7 +536,7 @@ TEST(scheduling, pos_mul_dense) {
   C.assemble();
   C.compute();
 
-  Tensor<double> expected("expected", {8}, {Dense});
+  Tensor<double> expected("expected", {8}, Format({Dense}));
   expected(i) = A(i) * B(i);
   expected.compile();
   expected.assemble();
@@ -545,9 +545,9 @@ TEST(scheduling, pos_mul_dense) {
 }
 
 TEST(scheduling, pos_mul_sparse) {
-  Tensor<double> A("A", {8}, {Sparse});
-  Tensor<double> B("B", {8}, {Sparse});
-  Tensor<double> C("C", {8}, {Dense});
+  Tensor<double> A("A", {8}, Format({Sparse}));
+  Tensor<double> B("B", {8}, Format({Sparse}));
+  Tensor<double> C("C", {8}, Format({Dense}));
 
   for (int i = 0; i < 8; i++) {
     if (i % 2 == 0) {
@@ -575,7 +575,7 @@ TEST(scheduling, pos_mul_sparse) {
   C.assemble();
   C.compute();
 
-  Tensor<double> expected("expected", {8}, {Dense});
+  Tensor<double> expected("expected", {8}, Format({Dense}));
   expected(i) = A(i) * B(i);
   expected.compile();
   expected.assemble();
@@ -584,9 +584,9 @@ TEST(scheduling, pos_mul_sparse) {
 }
 
 TEST(scheduling, pos_mul_dense_split) {
-  Tensor<double> A("A", {8}, {Sparse});
-  Tensor<double> B("B", {8}, {Dense});
-  Tensor<double> C("C", {8}, {Dense});
+  Tensor<double> A("A", {8}, Format({Sparse}));
+  Tensor<double> B("B", {8}, Format({Dense}));
+  Tensor<double> C("C", {8}, Format({Dense}));
 
   for (int i = 0; i < 8; i++) {
     if (i % 2 == 0) {
@@ -612,7 +612,7 @@ TEST(scheduling, pos_mul_dense_split) {
   C.assemble();
   C.compute();
 
-  Tensor<double> expected("expected", {8}, {Dense});
+  Tensor<double> expected("expected", {8}, Format({Dense}));
   expected(i) = A(i) * B(i);
   expected.compile();
   expected.assemble();
@@ -621,9 +621,9 @@ TEST(scheduling, pos_mul_dense_split) {
 }
 
 TEST(scheduling, pos_tile_coord_and_pos) {
-  Tensor<double> A("A", {8}, {Sparse});
-  Tensor<double> B("B", {8}, {Dense});
-  Tensor<double> C("C", {8}, {Dense});
+  Tensor<double> A("A", {8}, Format({Sparse}));
+  Tensor<double> B("B", {8}, Format({Dense}));
+  Tensor<double> C("C", {8}, Format({Dense}));
 
   for (int i = 0; i < 8; i++) {
     if (i % 2 == 0) {
@@ -650,7 +650,7 @@ TEST(scheduling, pos_tile_coord_and_pos) {
   C.assemble();
   C.compute();
 
-  Tensor<double> expected("expected", {8}, {Dense});
+  Tensor<double> expected("expected", {8}, Format({Dense}));
   expected(i) = A(i) * B(i);
   expected.compile();
   expected.assemble();
@@ -671,8 +671,8 @@ TEST(scheduling, spmv_warp_per_row) {
   const int iSIZE = 1024;
   const int jSIZE = 1024;
   Tensor<double> A("A", {iSIZE, jSIZE}, CSR);
-  Tensor<double> x("x", {jSIZE}, {Dense});
-  Tensor<double> y("y", {iSIZE}, {Dense});
+  Tensor<double> x("x", {jSIZE}, Format({Dense}));
+  Tensor<double> y("y", {iSIZE}, Format({Dense}));
 
   for (int i = 0; i < iSIZE; i++) {
     for (int j = 0; j < jSIZE; j++) {
@@ -709,7 +709,7 @@ TEST(scheduling, spmv_warp_per_row) {
   y.assemble();
   y.compute();
 
-  Tensor<double> expected("expected", {iSIZE}, {Dense});
+  Tensor<double> expected("expected", {iSIZE}, Format({Dense}));
   expected(i) = A(i, j) * x(j);
   stmt = expected.getAssignment().concretize();
   expected.compile(stmt);
@@ -719,8 +719,8 @@ TEST(scheduling, spmv_warp_per_row) {
 }
 
 TEST(scheduling, dense_pos_error) {
-  Tensor<double> x("x", {8}, {Dense});
-  Tensor<double> y("y", {8}, {Dense});
+  Tensor<double> x("x", {8}, Format({Dense}));
+  Tensor<double> y("y", {8}, Format({Dense}));
   IndexVar i("i"), ipos("ipos");
   y(i) = x(i);
 
@@ -733,8 +733,8 @@ TEST(scheduling, dense_pos_error) {
 }
 
 TEST(scheduling, pos_var_not_in_access) {
-  Tensor<double> x("x", {8}, {Dense});
-  Tensor<double> y("y", {8}, {Dense});
+  Tensor<double> x("x", {8}, Format({Dense}));
+  Tensor<double> y("y", {8}, Format({Dense}));
   IndexVar i("i"), ipos("ipos"), j("j");
   y(i) = x(i);
 
@@ -747,8 +747,8 @@ TEST(scheduling, pos_var_not_in_access) {
 }
 
 TEST(scheduling, pos_wrong_access) {
-  Tensor<double> x("x", {8}, {Dense});
-  Tensor<double> y("y", {8}, {Dense});
+  Tensor<double> x("x", {8}, Format({Dense}));
+  Tensor<double> y("y", {8}, Format({Dense}));
   IndexVar i("i"), ipos("ipos"), j("j");
   y(i) = x(i);
 
@@ -773,8 +773,8 @@ TEST(scheduling_eval_test, spmv_fuse) {
   int NNZ_PER_WARP = NNZ_PER_THREAD * WARP_SIZE;
   int NNZ_PER_TB = NNZ_PER_THREAD * BLOCK_SIZE;
   Tensor<double> A("A", {NUM_I, NUM_J}, CSR);
-  Tensor<double> x("x", {NUM_J}, {Dense});
-  Tensor<double> y("y", {NUM_I}, {Dense});
+  Tensor<double> x("x", {NUM_J}, Format({Dense}));
+  Tensor<double> y("y", {NUM_I}, Format({Dense}));
 
   srand(59393);
   for (int i = 0; i < NUM_I; i++) {
@@ -816,7 +816,7 @@ TEST(scheduling_eval_test, spmv_fuse) {
   y.assemble();
   y.compute();
 
-  Tensor<double> expected("expected", {NUM_I}, {Dense});
+  Tensor<double> expected("expected", {NUM_I}, Format({Dense}));
   expected(i) = A(i, j) * x(j);
   expected.compile();
   expected.assemble();

